org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/BreakpointManager.java

```diff
@@ -0,0 +1,166 @@
+/*******************************************************************************
+* Copyright (c) 2017 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.adapter;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+import org.eclipse.jdt.ls.debug.IBreakpoint;
+import org.eclipse.jdt.ls.debug.internal.Logger;
+
+public class BreakpointManager {
+    /**
+     * A collection of breakpoints registered with this manager.
+     */
+    private List<IBreakpoint> breakpoints;
+    private HashMap<String, HashMap<String, IBreakpoint>> sourceToBreakpoints;
+
+    /**
+     * Constructor.
+     */
+    public BreakpointManager() {
+        this.breakpoints = Collections.synchronizedList(new ArrayList<>(5));
+        this.sourceToBreakpoints = new HashMap<>();
+    }
+
+    /**
+     * Adds breakpoints to breakpoint manager.
+     * Deletes all breakpoints that are no longer listed.
+     * @param source
+     *              source path of breakpoints
+     * @param breakpoints
+     *              full list of breakpoints that locates in this source file
+     * @return the full breakpoint list that locates in the source file
+     */
+    public IBreakpoint[] addBreakpoints(String source, IBreakpoint[] breakpoints) {

```

Comment:
addBreakpoints is confusing since it replaces all the existing breakpoints to parameter array